### PR TITLE
Implement circuit for `EXP` OOG error

### PR DESCRIPTION
**Pull request by silathdiir**
### Description

Add `ErrorOOGExpGadget` to handle OOG error of `EXP` opcode.

Gas calculation could reference [this go-ethereum function](https://github.com/ethereum/go-ethereum/blob/master/core/vm/gas_table.go#L351).

Spec Markdown PR https://github.com/privacy-scaling-explorations/zkevm-specs/pull/388

### Issue Link

Close https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1231

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Tested with unit-test cases.

### CODE REVIEW

File: eth-types/src/evm_types.rs [hunk 1]
```diff
@@ -136,6 +136,10 @@ impl GasCost {
     pub const MEMORY_EXPANSION_LINEAR_COEFF: Self = Self(3);
     /// Constant gas for LOG[0-4] op codes
     pub const LOG: Self = Self(375);
+    /// Constant gas for LOG[0-4] op codes
```
Comment (roynalnaruto) [inline eth-types/src/evm_types.rs:139]: This line might have sneaked in by mistake?

File: zkevm-circuits/src/evm_circuit/execution/error_oog_exp.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    #[test]
+    fn test_oog_exp() {
+        test_ok(U256::zero());
+        test_ok(U256::one());
+        test_ok(1023.into());
+        test_ok(U256::MAX);
+        test_ok(rand_word());
+    }
+
+    fn test_ok(exponent: U256) {
```
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_exp.rs:189]: Might be good to also create a test scenario for an internal call that goes OOG. To check for the context restoration

**Review by roynalnaruto — COMMENTED** [review]
Overall LGTM! How about we also test the internal call scenario? As I see we only test OOG in the case of `is_root == 1`

**Review by roynalnaruto — APPROVED** [review]
Looks good!

**Review by lispc — APPROVED** [review]

**Pull request merged**